doc/portgroup.7
'\" t
.TH "PORTGROUP" "7" "2\&.11\&.99" "MacPorts 2\&.11\&.99" "MacPorts Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
portgroup \- MacPorts PortGroup command reference
.SH "SYNOPSIS"
.sp
.nf
PortGroup gnustep 1\&.0
PortGroup perl5 1\&.0
PortGroup python24 1\&.0
PortGroup python25 1\&.0
PortGroup python30 1\&.0
PortGroup ruby 1\&.0
PortGroup xcode 1\&.0
.fi
.SH "DESCRIPTION"
.sp
MacPorts defines the notion of PortGroup classes with the PortGroup command\&. These classes factorize the work for similar ports\&. Port maintainers willing to use PortGroup need to have some notion of Portfiles\&. See \fBportfile\fR(7)\&.
.SH "XCODE PORTGROUP"
.sp
\fBPortGroup xcode\fR is here to easily port Xcode\-based opensource software\&. It handles configuration, build and destroot phases\&. It also defines some values for Xcode\-based software\&. A minimum Portfile using the \fBxcode PortGroup\fR class only defines the fetch and the checksum phases\&.
.sp
Using \fBPortGroup xcode\fR is a way to make your port more robust to Xcode version updates as the PortGroup is tested against all supported macOS and Xcode versions\&.
.SS "XCODE PORTGROUP SUGAR"
.sp
Portfiles using \fBxcode PortGroup\fR do not need to define the following variables:
.PP
categories
.RS 4
\fBDefault:\fR
aqua
.RE
.PP
platforms
.RS 4
\fBDefault:\fR
macosx
.RE
.PP
use_configure
.RS 4
\fBDefault:\fR
no
.RE
.sp
Moreover, Portfiles using \fBPortGroup xcode\fR do not need to define any build or destroot phase\&. Some options are provided for tailoring these phases to what the software requires\&.
.SS "XCODE PORTGROUP SPECIFIC VARIABLES"
.sp
Portfiles using \fBxcode PortGroup\fR may define the following variables:
.PP
xcode\&.project
.RS 4
Path (relative to ${build\&.dir} and ${destroot\&.dir}) of the Xcode project\&. If unset, let xcode tool figure it out\&. It usually succeeds if there is a single project in the directory\&.
.TS
tab(:);
lt lt.
T{
\fBType:\fR
T}:T{
optional
T}
.TE
.sp 1
\fBExample:\fR
.sp
.if n \{\
.RS 4
.\}
.nf
xcode\&.project ${name}\&.xcode
.fi
.if n \{\
.RE
.\}
.RE
.PP
xcode\&.configuration
.RS 4
Project configuration/buildstyle to use\&.
.TS
tab(:);
lt lt
lt lt.
T{
\fBType:\fR
T}:T{
optional
T}
T{
\fBDefault:\fR
T}:T{
Deployment
T}
.TE
.sp 1
\fBExample:\fR
.sp
.if n \{\
.RS 4
.\}
.nf
xcode\&.configuration Main
.fi
.if n \{\
.RE
.\}
.RE
.PP
xcode\&.target
.RS 4
If present, overrides build\&.target and destroot\&.target\&.
.TS
tab(:);
lt lt.
T{
\fBType:\fR
T}:T{
optional
T}
.TE
.sp 1
.RE
.PP
xcode\&.build\&.settings
.RS 4
Additional settings passed to the xcode build tool during the build phase\&. These settings should be in the X=Y form\&.
.TS
tab(:);
lt lt.
T{
\fBType:\fR
T}:T{
optional
T}
.TE
.sp 1
\fBExample:\fR
.sp
.if n \{\
.RS 4
.\}
.nf
xcode\&.build\&.settings FRAMEWORK_SEARCH_PATHS=${prefix}/Library/Frameworks
.fi
.if n \{\
.RE
.\}
.RE
.PP
xcode\&.destroot\&.type
.RS 4
Type of project that will be installed\&. This tells Xcode PortGroup how to destroot the project\&. Correct values are application and framework\&.
.TS
tab(:);
lt lt
lt lt.
T{
\fBType:\fR
T}:T{
optional
T}
T{
\fBDefault:\fR
T}:T{
application
T}
.TE
.sp 1
\fBExample:\fR
.sp
.if n \{\
.RS 4
.\}
.nf
xcode\&.destroot\&.type framework
.fi
.if n \{\
.RE
.\}
.RE
.PP
xcode\&.destroot\&.path
.RS 4
Where to install the build product\&.
.TS
tab(:);
lt lt
lt lt.
T{
\fBType:\fR
T}:T{
optional
T}
T{
\fBDefault:\fR
T}:T{
${prefix}/Library/Frameworks or /Applications/MacPorts depending on xcode\&.destroot\&.type
T}
.TE
.sp 1
.RE
.PP
xcode\&.destroot\&.settings
.RS 4
Additional settings passed to the xcode build tool during the destroot phase\&. These settings should be in the X=Y form\&.
.TS
tab(:);
lt lt.
T{
\fBType:\fR
T}:T{
optional
T}
.TE
.sp 1
\fBExample:\fR
.sp
.if n \{\
.RS 4
.\}
.nf
xcode\&.destroot\&.settings SKIP_INSTALL=NO
.fi
.if n \{\
.RE
.\}
.RE
.SS "STANDARD PORTFILE VARIABLES AFFECTING XCODE PORTGROUP"
.sp
The following standard Portfile variables actually affect \fBxcode PortGroup\fR\&. In most cases, you will not need to set any of these variables in the Portfile\&.
.PP
destroot
.RS 4
Where to destroot the project\&.
.RE
.PP
build\&.cmd
.RS 4
The command to invoke to build the project\&.
\fBxcode PortGroup\fR
sets this variable to ${xcodebuildcmd}\&.
.RE
.PP
build\&.target
.RS 4
The Xcode target to build\&.
\fBxcode PortGroup\fR
sets this variable to "", thus building all targets\&. This variable will be ignored if xcode\&.target is set\&.
.RE
.PP
build\&.pre_args
.RS 4
Additional parameters for ${build\&.cmd}\&.
.TS
tab(:);
lt lt.
T{
\fBDefault:\fR
T}:T{
none
T}
.TE
.sp 1
.RE
.PP
build\&.args
.RS 4
Main parameters for ${build\&.cmd}\&.
.TS
tab(:);
lt lt.
T{
\fBDefault:\fR
T}:T{
build
T}
.TE
.sp 1
.RE
.PP
build\&.post_args
.RS 4
Additional parameters for ${build\&.cmd}\&.
.TS
tab(:);
lt lt.
T{
\fBDefault:\fR
T}:T{
none
T}
.TE
.sp 1
.RE
.PP
build\&.dir
.RS 4
Where to build the project from (i\&.e\&. where the Xcode project is)\&.
.TS
tab(:);
lt lt.
T{
\fBDefault:\fR
T}:T{
${worksrcpath}
T}
.TE
.sp 1
.RE
.PP
destroot\&.cmd
.RS 4
The command to invoke to destroot the project\&.
\fBxcode PortGroup\fR
sets this variable to xcodebuildcmd\&.
.RE
.PP
destroot\&.target
.RS 4
The Xcode target to install\&.
\fBxcode PortGroup\fR
sets this variable to "", thus installing all targets\&. This variable will be ignored if xcode\&.target is set\&.
.RE
.PP
destroot\&.pre_args
.RS 4
Additional parameters for ${destroot\&.cmd}\&.
.TS
tab(:);
lt lt.
T{
\fBDefault:\fR
T}:T{
none
T}
.TE
.sp 1
.RE
.PP
destroot\&.args
.RS 4
Main parameters for ${destroot\&.cmd}\&.
.TS
tab(:);
lt lt.
T{
\fBDefault:\fR
T}:T{
install
T}
.TE
.sp 1
.RE
.PP
destroot\&.post_args
.RS 4
Additional parameters for ${destroot\&.cmd}\&.
.TS
tab(:);
lt lt.
T{
\fBDefault:\fR
T}:T{
none
T}
.TE
.sp 1
.RE
.PP
destroot\&.dir
.RS 4
Where to destroot the project from (i\&.e\&. where the Xcode project is)\&.
.TS
tab(:);
lt lt.
T{
\fBDefault:\fR
T}:T{
${worksrcpath}
T}
.TE
.sp 1
.RE
.SH "GNUSTEP PORTGROUP"
.sp
\fBPortGroup gnustep\fR is here to easily port GNUstep\-based opensource software using the GNU objective\- C runtime\&. It handles configuration, build and destroot phases\&. It also defines some values for GNUstep\-based software\&. A minimum Portfile using the \fBgnustep PortGroup\fR class only defines the fetch and the checksum phases\&.
.SS "GNUSTEP FILESYSTEM LAYOUTS"
.sp
PortGroup gnustep also supports both the traditional gnustep file layout and the new fhs file layout\&. However, the ports themselves do not necessarily support both\&. The Portfiles have access to many procedures in dealing with these two layouts:
.PP
set_gnustep_make
.RS 4
Sets GNUSTEP_MAKEFILES according to the FilesystemLayout
.RE
.PP
set_gnustep_env
.RS 4
Sets DYLD_LIBRARY_PATH and PATH for the gnustep FilesystemLayout
.RE
.PP
gnustep_layout
.RS 4
Returns true (1) if current file layout is gnustep
.RE
.PP
set_system_library
.RS 4
Sets GNUSTEP_SYSTEM_LIBRARY according to the FilesystemLayout
.RE
.PP
set_local_library
.RS 4
Sets GNUSTEP_LOCAL_LIBRARY according to the FilesystemLayout
.RE
.SS "GNUSTEP PORTGROUP SUGAR"
.sp
Portfiles using \fBgnustep PortGroup\fR do not need to define the following variables:
.PP
categories
.RS 4
\fBDefault:\fR
gnustep
.RE
.PP
homepage
.RS 4
\fBDefault:\fR
\m[blue]\fBhttp://www\&.gnustep\&.org/\fR\m[]
.RE
.PP
master_sites
.RS 4
\fBDefault:\fR
gnustep:core
.RE
.PP
depends_lib
.RS 4
\fBDefault:\fR
gnustep\-core
.RE
.PP
use_configure
.RS 4
\fBDefault:\fR
no
.RE
.PP
configure\&.env
.RS 4
\fBDefault:\fR
DYLD_LIBRARY_PATH PATH
.RE
.PP
configure\&.pre_args\-append
.RS 4
\fBDefault:\fR
CC=gcc\-mp\-4\&.2 GNUSTEP_MAKEFILES
.RE
.PP
build\&.type
.RS 4
\fBDefault:\fR
gnu
.RE
.PP
build\&.env
.RS 4
\fBDefault:\fR
DYLD_LIBRARY_PATH PATH
.RE
.PP
build\&.pre_args\-append
.RS 4
\fBDefault:\fR
messages=yes
.RE
.PP
destroot\&.env
.RS 4
\fBDefault:\fR
DYLD_LIBRARY_PATH PATH
.RE
.PP
destroot\&.pre_args\-append
.RS 4
\fBDefault:\fR
messages=yes
.RE
.sp
Moreover, Portfiles using \fBPortGroup gnustep\fR do not need to define any build or destroot phase\&. Some options are provided for tailoring these phases to what the software requires\&. A mechanism is also provided to ease the patch process\&.
.SS "GNUSTEP PORTGROUP SPECIFIC VARIABLES"
.sp
Portfiles using \fBgnustep PortGroup\fR may define the following variables:
.PP
gnustep\&.post_flags
.RS 4
an associative array which specifies the sub\-directories relative to ${worksrcpath} and the SHARED_LD_POSTFLAGS variables to be added to GNUmakefile\&.preamble in those sub\-directories\&. This helps making the patching process easier on Darwin\&.
.TS
tab(:);
lt lt.
T{
\fBType:\fR
T}:T{
optional
T}
.TE
.sp 1
\fBExample:\fR
.sp
.if n \{\
.RS 4
.\}
.nf
platform darwin {
    array set gnustep\&.post_flags {
        BundleSubDir "\-lfoo \-lbar"
    }
}
.fi
.if n \{\
.RE
.\}
.RE
.PP
gnustep\&.cc
.RS 4
.TS
tab(:);
lt lt
lt lt.
T{
\fBType:\fR
T}:T{
optional
T}
T{
\fBDefault:\fR
T}:T{
gcc\-mp\-4\&.2
T}
.TE
.sp 1
\fBExample:\fR
.sp
.if n \{\
.RS 4
.\}
.nf
gnustep\&.cc gcc\-mp\-4\&.3
.fi
.if n \{\
.RE
.\}
.RE
.PP
variant with_docs
.RS 4
Many GNUstep packages include a Documentation sub\-directory that is not built by default\&. Enabling this variant builds and installs the included documentation\&.
.TS
tab(:);
lt lt.
T{
\fBType:\fR
T}:T{
optional
T}
.TE
.sp 1
\fBExample:\fR
.sp
.if n \{\
.RS 4
.\}
.nf
port install gnustep\-gui +with_docs
.fi
.if n \{\
.RE
.\}
.RE
.SH "SEE ALSO"
.sp
\fBport\fR(1), \fBmacports.conf\fR(5), \fBportfile\fR(7), \fBportstyle\fR(7), \fBporthier\fR(7)
.SH "AUTHORS"
.sp
.if n \{\
.RS 4
.\}
.nf
(C) 2013 The MacPorts Project
Paul Guyot <pguyot@kallisys\&.net>
Yves de Champlain <yves@macports\&.org>
Rainer Mueller <raimue@macports\&.org>
.fi
.if n \{\
.RE
.\}


NEWS
== Overview of changes from MacPorts 2.0.4 to 2.1.0 ==

# For regular users:

*) New 'rev-upgrade' action that checks for and rebuilds ports that may have
   become incorrectly linked when a dependency was upgraded to a new,
   binary-incompatible version. This runs automatically after upgrades and
   installs by default, and can be turned off or configured to only report
   problems in macports.conf.

*) The pkg and mpkg targets will create flat packages on 10.6+.

*) Build-time dependencies will no longer be installed if a prebuilt
   archive is available for download (and port is configured to use it).

*) Config file archive_sites.conf can be used to download archives from
   custom locations.

*) 'buildfromsource' macports.conf option allows permanently choosing
   source-only or binary-only mode.

*) New macports.conf options 'host_blacklist' and 'preferred_hosts', for
   controlling which download hosts will be used.

# For Portfile writers:

*) New -W option for reinplace that sets a base directory for all file
   arguments.

*) Livecheck can now find multiple regex matches on each line.

*) license_noconflict option can be used to override automatic license
   compatibility checks.

*) compiler.whitelist, compiler.blacklist, and compiler.fallback options allow
   simpler, detailed control over which compilers are used to build ports.


== Overview of changes from MacPorts 1.9.2 to 2.0.0 ==

# For regular users:

*) Downloaded archives are now checked for integrity using digital signatures.
   New pubkeys.conf file allows configuring keys to trust.

*) Actions can be abbreviated to an unambiguous prefix of the full name.

*) Ports that create StartupItems (aka launchd plists) should now be
   installable by unprivileged users.

*) Port images are now stored as archives. Archive mode is now effectively
   always on. Direct mode and the flat registry format are no longer
   available. Conversion will happen automatically.

# For Portfile writers:

*) Multiple ports can now be defined in one Portfile using the 'subport'
   procedure.

*) New option 'add_users' can be set to ensure that the listed users are
   created at the right times during the install process.

*) SHA256 checksums are now supported.

== Overview of changes from MacPorts 1.8.2 to 1.9.0 ==

# For regular users:

*) All output is now logged to a file, whose location will be shown if a port
   target fails. This should be attached to bug reports. The new 'log' action
   can also be used to view the log for a port.
*) Command line options can be abbreviated if unambiguous.
*) New actions: rdeps, rdependents, setrequested, unsetrequested
*) --follow-dependencies option for uninstall (ports with the requested flag
   set or with other dependents are skipped).
*) New pseudo-ports and selectors: requested, unrequested, leaves, depof:,
   rdepof:, dependentof:, rdependentof:.
*) Architectures of dependencies are checked before install, and in many cases
   will be automatically fixed by upgrading with +universal.
*) New sqlite registry format which is faster and lacks some of the old
   format's inconvenient limitations.
*) Explicitly deselected variants (e.g. -foo) are now recorded, so they will
   not be set to + when upgrading if +foo appears in variants.conf or
   the port's default_variants.

# For Portfile writers:

*) New os.subplatform variable.
*) New configure.ld_archflags option to match configure.cc_archflags et al.
*) 'platform' acts like an 'if' statement rather than a variant.
*) New supported_archs option for indicating a port only supports certain
   archs or is noarch.
*) pre-/post-deactivate and pre-/post-uninstall procedures can now be used,
   and pre-/post-activate procedures are run every time the port is activated,
   not just the first time when it is installed.

== Overview of changes from MacPorts 1.7.1 to 1.8.0 ==

# For regular users:

*) Compatible with Mac OS X 10.6 Snow Leopard, requires at least 10.4 Tiger
*) CPU architecture to build for is now configurable
*) Performance and memory usage improvements
*) New -y option for dry-run
*) 'port search' has many new options that make it more flexible
*) 'port upgrade' now follows the right dependencies when variants change them
*) 'port info' and 'port variants' indicate selected and default variants;
   'port variants' also shows conflicts and dependencies between variants
*) A port can be marked as being replaced by another port; upgrading will
   perform the replacement
*) upgrade accepts new options: --force, --enforce-variants, --no-replace
*) Ports can be marked as conflicting, and conflicts will be detected before
   the install process starts
*) Proxy information can now be obtained from System Preferences and from
   macports.conf
*) Activating a port will now first deactivate any other version of the same
   port that is already active
*) MacPorts can be configured to drop root privileges when not needed
*) port(1) will exit with an appropriate status code when an error is
   encountered (use -p to continue despite errors)

# For Portfile writers:

*) New options: replaced_by, conflicts, depends_fetch, depends_extract, license
*) build_arch variable indicates which architecture should be targeted. Flags
   used to implement this are configured per-compiler, for example,
   configure.cc_archflags will be passed to configure.cc in CFLAGS
*) Dependencies can be satisfied by ports that are installed but not
   available in the ports tree
*) Deleting all the contents of an option no longer unsets the option entirely
*) Using an alternative fetch or extract type now adds appropriate
   dependencies automatically, e.g. 'use_zip yes' adds a dep on unzip, and
   'fetch.type svn' adds a dep on subversion
*) Running phases prior to install should now work reliably when the port is
   already installed

== Overview of changes from MacPorts 1.6.0 to 1.7.0 ==

# For regular users:


src/registry2.0/portgroup.c
/*
 * vim:tw=80:expandtab
 *
 * Copyright (c) 2014 The MacPorts Project
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#if HAVE_CONFIG_H
#include <config.h>
#endif

#include <sqlite3.h>
#include <stdlib.h>
#include <string.h>
#include <tcl.h>

#include <cregistry/portgroup.h>
#include <cregistry/util.h>

#include "portgroup.h"
#include "portgroupobj.h"
#include "registry.h"
#include "util.h"

/**
 * Converts a command name into a `reg_portgroup`.
 *
 * @param [in] interp  Tcl interpreter to check within
 * @param [in] name    name of portgroup to get
 * @param [out] errPtr description of error if the portgroup can't be found
 * @return             a portgroup, or NULL if one couldn't be found
 * @see get_object
 */
static reg_portgroup* get_portgroup(Tcl_Interp* interp, char* name, reg_error* errPtr) {
    return (reg_portgroup*)get_object(interp, name, "portgroup", portgroup_obj_cmd, errPtr);
}

/**
 * Removes the portgroup from the Tcl interpreter. Doesn't actually delete it since
 * that's the registry's job. This is written to be used as the
 * `Tcl_CmdDeleteProc` for an portgroup object command.
 *
 * @param [in] clientData address of a reg_portgroup to remove
 */
void delete_portgroup(ClientData clientData) {
    reg_portgroup* portgroup = (reg_portgroup*)clientData;
    free(portgroup->proc);
    portgroup->proc = NULL;
}

/**
 * registry::portgroup open id name version size sha256
 *
 * Opens a portgroup matching the given parameters.
 */
static int portgroup_open(Tcl_Interp* interp, int objc, Tcl_Obj* const objv[]) {
	reg_registry* reg = registry_for(interp, reg_attached);
	if (objc != 7) {
		Tcl_WrongNumArgs(interp, 1, objv, "open id name version size sha256");
		return TCL_ERROR;
	} else if (reg == NULL) {
		return TCL_ERROR;
	} else {
		char* id = Tcl_GetString(objv[2]);
		char* name = Tcl_GetString(objv[3]);
        char* version = Tcl_GetString(objv[4]);
        char* size = Tcl_GetString(objv[5]);
        char* sha256 = Tcl_GetString(objv[6]);

		reg_error error;
		reg_portgroup* portgroup = reg_portgroup_open(reg, id, name, version, size, sha256, &error);
		if (portgroup != NULL) {
			Tcl_Obj* result;
			if (portgroup_to_obj(interp, &result, portgroup, NULL, &error)) {
				Tcl_SetObjResult(interp, result);
				return TCL_OK;
			}
		}
		return registry_failed(interp, &error);
	}
	return TCL_ERROR;
}

/**
 * registry::portgroup close portgroup
 *
 * Closes a portgroup. It will remain in the registry.
 */
static int portgroup_close(Tcl_Interp* interp, int objc, Tcl_Obj* const objv[]) {
	if (objc != 3) {
		Tcl_WrongNumArgs(interp, 1, objv, "close portgroup");
		return TCL_ERROR;
	} else {
		reg_error error;
		char* proc = Tcl_GetString(objv[2]);
		reg_portgroup* portgroup = get_portgroup(interp, proc, &error);
		if (portgroup == NULL) {
			return registry_failed(interp, &error);
		} else {
			Tcl_DeleteCommand(interp, proc);
			return TCL_OK;
		}
	}
}

typedef struct {
    char* name;
    reg_strategy strategy;
} strategy_type;

static strategy_type strategies[] = {
    { "-exact",  reg_strategy_exact },
    { "-glob",   reg_strategy_glob },
    { "-regexp", reg_strategy_regexp },
    { "-null",   reg_strategy_null },
    { "--",      reg_strategy_exact },
    { NULL, 0 }
};

/*
 * registry::portgroup search ?key value ...?
 *
 * Searches the registry for portgroups for which each key's value is equal to the
 * given value. To find all portgroups, call `portgroup search` with no key-value pairs.
 * For each key, can be given an option of -exact, -glob, -regexp or -null to
 * specify the matching strategy; defaults to exact.
 */
static int portgroup_search(Tcl_Interp* interp, int objc, Tcl_Obj* const objv[]) {
    int i, j;
    reg_registry* reg = registry_for(interp, reg_attached);
    if (reg == NULL) {
        return TCL_ERROR;
    } else {
        char** keys;
        char** vals;
        int* strats;
        int key_count = 0;
        reg_portgroup** portgroups;
        reg_error error;
        int portgroup_count;
        for (i = 2; i < objc;) {
            int index, strat_index, val_length;
            if (Tcl_GetIndexFromObj(interp, objv[i], portgroup_props, "search key",
                        0, &index) != TCL_OK) {
                return TCL_ERROR;
            }

            /* we ate the key value */
            i++;

            /* check whether there's a strategy */
            if (Tcl_GetString(objv[i])[0] == '-'
                    && Tcl_GetIndexFromObjStruct(interp, objv[i], strategies,
                        sizeof(strategy_type), "option", 0, &strat_index)
                    != TCL_ERROR) {
                /* this key has a strategy specified, eat the strategy parameter */
                i++;

                if (strategies[strat_index].strategy != reg_strategy_null) {
                    /* this key must also have a value */

                    if (Tcl_GetStringFromObj(objv[i], &val_length) == NULL
                            || val_length == 0) {
                        Tcl_WrongNumArgs(interp, 2, objv,
                                "search ?key ?options? value ...?");
                        return TCL_ERROR;
                    }

                    i++;
                }
            } else {
                /* this key must also have a value */

                if (Tcl_GetStringFromObj(objv[i], &val_length) == NULL
                        || val_length == 0) {
                    Tcl_WrongNumArgs(interp, 2, objv,
                            "search ?key ?options? value ...?");
                    return TCL_ERROR;
                }

                i++;
            }

            key_count++;
        }

        keys = malloc(key_count * sizeof(char*));
        vals = malloc(key_count * sizeof(char*));
        strats = malloc(key_count * sizeof(int));
        if (!keys || !vals || !strats) {
            if (keys) {
                free(keys);
            }
            if (vals) {
                free(vals);
            }
            if (strats) {
                free(strats);
            }
            return TCL_ERROR;
        }
        for (i = 2, j = 0; i < objc && j < key_count; j++) {
            int strat_index;

            keys[j] = Tcl_GetString(objv[i++]);

            /* try to get the strategy */
            if (Tcl_GetString(objv[i])[0] == '-'
                    && Tcl_GetIndexFromObjStruct(interp, objv[i], strategies,
                        sizeof(strategy_type), "option", 0, &strat_index)
                    != TCL_ERROR) {
                /* this key has a strategy specified */
                i++;

                strats[j] = strategies[strat_index].strategy;
            } else {
                /* use default strategy */
                strats[j] = reg_strategy_exact;
            }

            if (strats[j] != reg_strategy_null) {
                vals[j] = Tcl_GetString(objv[i++]);
            } else {
                vals[j] = NULL;
            }
        }
        portgroup_count = reg_portgroup_search(reg, keys, vals, strats, key_count,
                &portgroups, &error);
        free(keys);
        free(vals);
        free(strats);
        if (portgroup_count >= 0) {
            int retval;
            Tcl_Obj* resultObj;
            Tcl_Obj** objs;
            if (list_portgroup_to_obj(interp, &objs, portgroups, portgroup_count, &error)){
                resultObj = Tcl_NewListObj(portgroup_count, objs);
                Tcl_SetObjResult(interp, resultObj);
                free(objs);
                retval = TCL_OK;
            } else {
                retval = registry_failed(interp, &error);
            }
            free(portgroups);
            return retval;
        }
        return registry_failed(interp, &error);
    }
}

typedef struct {
    char* name;
    int (*function)(Tcl_Interp* interp, int objc, Tcl_Obj* const objv[]);
} portgroup_cmd_type;

static portgroup_cmd_type portgroup_cmds[] = {
    /* Global commands */
    { "open", portgroup_open },
    { "close", portgroup_close },
    { "search", portgroup_search },
    { NULL, NULL }
};

/*
 * registry::portgroup cmd ?arg ...?
 *
 * Commands manipulating portgroup entries in the registry. This can be called `registry::portgroup`
 */
int portgroup_cmd(ClientData clientData UNUSED, Tcl_Interp* interp, int objc,
        Tcl_Obj* const objv[]) {
    int cmd_index;
    if (objc < 2) {
        Tcl_WrongNumArgs(interp, 1, objv, "cmd ?arg ...?");
        return TCL_ERROR;
    }
    if (Tcl_GetIndexFromObjStruct(interp, objv[1], portgroup_cmds,
                sizeof(portgroup_cmd_type), "cmd", 0, &cmd_index) == TCL_OK) {
        portgroup_cmd_type* cmd = &portgroup_cmds[cmd_index];
        return cmd->function(interp, objc, objv);
    }
    return TCL_ERROR;
}
